design/m1_push_to_file_design.md

```diff
@@ -0,0 +1,1247 @@
+- Add new documentation to describe the push-to-file annotation configuration
+- Add documentation for the process to upgrade from the Secrets Provider legacy (environment variable based) configuration to the annotation-based configuration.
+
+## Version update
+
+The M1 Push-to-File feature will necessitate a minor version bump for
+the Secrets Provider. Because configuration will maintain backwards
+compatibility, it is not expected that a major version bump will be required.
+
+## Security
+
+### Secrets File Attributes
+
+Since the Push-to-File feature involves writing sensitive information to
+files that are shared by an application container, care needs to be taken
+so that files are created with the proper file attributes:
+
+- File permissions
+- File owner/UID
+- File group/GID
+
+By default, the Secrets Provider will create secrets files with the following
+file attributes, based on the default user defined in the Secrets Provider
+container image:
+
+- File permissions: 660         (`rw-rw----`)
+- File owner: secrets-provider  (UID 777)
+- File group: secrets-provider  (GID 777)
+
+If the application container is built to run with a non-root user that has
+different a UID and GID than the default values shown above, then the
+application will not be able to access secrets files that have been
+created using the default attributes.
+
+In these cases, users will want to include a
+[Pod Security Context](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#podsecuritycontext-v1-core)
+in their Pod Deployment to change either the user that is running inside
+the Secrets Provider container, or the GID with which secrets files will be
+created. Pod Security Context configuration can be added to a Pod Deployment
+as decribed [here](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/).
+
+For example, to configure the user and group with which the Secrets Provider
+container runs, and enforce that the Pod containers must run as a non-root
+user, the following can be included in the Pod spec:
+
+```
+    securityContext:
+      runAsUser: 1000
+      runAsGroup: 3000
+      runAsNonRoot: true
+```
+
+In this example, the access to secrets files can be restricted further by
+modifying the file permissions that are used when secrets files are created,
+for example, using the following annotation:
+
+```
+    conjur.org/secret-file-perms.database=600
+```
+
+Another option is to configure the GID with which all files (including secrets
+files) are created in `emptyDir` volumes:
+
+```
+    securityContext:
+      fsGroup: 2000
+```
+
+Note that defining an `fsGroup` will cause the file permissions bits to be
+OR'd with `rw-rw----`.
+
+(For OpenShift deployments, the `fsGroup` may need to be configured as `0`.)
```

Comment:
Does OpenShift even support this?  The reason we created a separate enterprise Conjur Follower was to avoid running as root, or specific user IDs.

Reply:
Reworded per our discussion to describe why this isn't a concern for OpenShift.